xpcPlugin/MessageHandlers.cpp

```diff
@@ -533,16 +533,23 @@ namespace XPC
 			connections[connectionKey] = connection;
 		}
 
-		unsigned char response[4096] = "RESP";
+		unsigned char response[XPC_MAX_MESSAGE_SIZE] = "RESP";
 		response[5] = drefCount;
 		std::size_t cur = 6;
 		for (int i = 0; i < drefCount; ++i)
 		{
-			float values[255];
-			int count = DataManager::Get(connection.getdRequest[i], values, 255);
+			float values[XPC_MAX_DREF_VALUES];
+			int count = DataManager::Get(connection.getdRequest[i], values, XPC_MAX_DREF_VALUES);
+			int dataSize = count * sizeof(float);
+			if (cur + drefCount - i + dataSize > XPC_MAX_MESSAGE_SIZE) {
+				Log::FormatLine(LOG_ERROR, "GETD", "ERROR: omitting data (%i bytes) for dref %i to avoid exceeding message buffer for connection %i.",
+					dataSize, i, connection.id);
+				count = 0;
+				dataSize = 0;
+			}
```

Comment:
this omits data if it would prevent writing at least a single count byte for all remaining drefs.  this will always succeed as long as the message size is larger than the max allowed dref count, e.g. XPC_MAX_MESSAGE_SIZE > 255 + message header.  

if there's concern about some clients not dealing with empty value arrays (I only tested the python client where it's fine), this could be changed to return only a single value from the array if there's insufficient space to write 1 byte + 1 float for all remaining drefs.